crates/transpiler/src/passes/asap_schedule_analysis.rs

```diff
@@ -0,0 +1,218 @@
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::passes::alap_schedule_analysis::TimeOps;
+use crate::TranspilerError;
+use hashbrown::HashMap;
+use pyo3::prelude::*;
+use pyo3::types::PyDict;
+use qiskit_circuit::dag_circuit::{DAGCircuit, Wire};
+use qiskit_circuit::dag_node::{DAGNode, DAGOpNode};
+use qiskit_circuit::operations::{OperationRef, StandardInstruction};
+use qiskit_circuit::{Clbit, Qubit};
+use rustworkx_core::petgraph::prelude::NodeIndex;
+
+pub fn run_asap_schedule_analysis<T: TimeOps>(
+    dag: &DAGCircuit,
+    clbit_write_latency: T,
+    node_durations: HashMap<NodeIndex, T>,
+) -> PyResult<HashMap<NodeIndex, T>> {
+    if dag.qregs().len() != 1 || !dag.qregs_data().contains_key("q") {
+        return Err(TranspilerError::new_err(
+            "ASAP schedule runs on physical circuits only",
+        ));
+    }
+
+    let mut node_start_time: HashMap<NodeIndex, T> = HashMap::new();
+    let mut idle_after: HashMap<Wire, T> = HashMap::new();
+
+    let zero = T::zero();
+
+    for index in 0..dag.qubits().len() {
+        idle_after.insert(Wire::Qubit(Qubit::new(index)), zero);
+    }
+
+    for index in 0..dag.clbits().len() {
+        idle_after.insert(Wire::Clbit(Clbit::new(index)), zero);
+    }
+
+    for node_index in dag.topological_op_nodes()? {
+        let op = dag[node_index].unwrap_operation();
+
+        let qargs: Vec<Wire> = dag
+            .qargs_interner()
+            .get(op.qubits)
+            .iter()
+            .map(|&q| Wire::Qubit(q))
+            .collect();
+        let cargs: Vec<Wire> = dag
+            .cargs_interner()
+            .get(op.clbits)
+            .iter()
+            .map(|&c| Wire::Clbit(c))
+            .collect();
+
+        let &op_duration = node_durations
+            .get(&node_index)
+            .ok_or_else(|| TranspilerError::new_err("No duration for node"))?;
+
+        let op_view = op.op.view();
+        let is_gate_or_delay = matches!(
+            op_view,
+            OperationRef::Gate(_)
+                | OperationRef::StandardGate(_)
+                | OperationRef::StandardInstruction(StandardInstruction::Delay(_))
+        );
+
+        // compute t0, t1: instruction interval, note that
+        // t0: start time of instruction
+        // t1: end time of instruction
+
+        let (t0, t1) = if is_gate_or_delay {
+            let &t0 = qargs
+                .iter()
+                .map(|q| idle_after.get(q).unwrap_or(&zero))
+                .fold(&zero, |acc, x| T::max(acc, x));
+            (t0, t0 + op_duration)
+        } else if matches!(
+            op_view,
+            OperationRef::StandardInstruction(StandardInstruction::Measure)
+        ) {
+            // Measure instruction handling is bit tricky due to clbit_write_latency
+            let t0q = qargs
+                .iter()
+                .map(|q| *idle_after.get(q).unwrap_or(&zero))
```

Comment:
Since these are guaranteed to exist, try using indexing instead of defaulting to zero. If something outside of there ever pops up we'd know that there is something wrong with the pass.
```suggestion
                .map(|q| *idle_after[q])
```